### PR TITLE
Dec 1064 update item metadata saving

### PR DESCRIPTION
**Pull request by jonhartzler**
Why: the current implementation was validating the metadata on create which was causing new item creation to fail also there was a bug with the custom metadata validator in that it always used the first loaded collection configuration to validate agains.
What:  I separated the routes for updating items and metadata.  All routes to items should only save item data.  There is a new api route v1/items/ID/metadata that saves the metadata.   There is also a new save_metadata class for saving metadata.  
SaveItem no longer saves metadata.

### CODE REVIEW

File: spec/services/save_metadata_spec.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  it "sets the metadata with the metadata setter" do
+    expect(Metadata::Setter).to receive(:call).with(item, params).and_return(true)
+    subject
+  end
+
+  it "does not set the metadata if the params are nil" do
+    expect(Metadata::Setter).to_not receive(:call).with(params)
+    params = nil
```
Comment (jonhartzler) [inline spec/services/save_metadata_spec.rb]: hound has this wrong because it does not know about the let statement ahead.  I am not planning on correcting this.

File: spec/services/save_metadata_spec.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    allow(Metadata::Setter).to receive(:call).and_return(true)
+  end
+
+  it "sets the metadata with the metadata setter" do
+    expect(Metadata::Setter).to receive(:call).with(item, params).and_return(true)
+    subject
+  end
+
+  it "does not set the metadata if the params are nil" do
+    expect(Metadata::Setter).to_not receive(:call).with(params)
```
Comment (jgondron) [inline spec/services/save_metadata_spec.rb]: Don't need the with(params)

File: spec/services/save_metadata_spec.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    params = nil
+    subject
+  end
+
+  it "calls item#save if the item_metadata is valid? " do
+    expect(item_metadata).to receive(:valid?).and_return(true)
```
Comment (jgondron) [inline spec/services/save_metadata_spec.rb]: Should just be an allow.

File: spec/services/save_metadata_spec.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    expect(item).to receive(:save).and_return(true)
+    subject
+  end
+
+  it "does not call item#save if the item_metadata is not valid?" do
+    expect(item_metadata).to receive(:valid?).and_return(false)
```
Comment (jgondron) [inline spec/services/save_metadata_spec.rb]: Should just be an allow.

File: spec/services/save_metadata_spec.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    expect(item).to_not receive(:save)
+    subject
+  end
+
+  it "returns the item_metadata if the process is successful" do
+    expect(subject).to eq(item_metadata)
+  end
+
+  it "returns false if the the item_metadata is not valid?" do
+    expect(item_metadata).to receive(:valid?).and_return(false)
```
Comment (jgondron) [inline spec/services/save_metadata_spec.rb]: Should just be an allow.

File: spec/services/save_metadata_spec.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    expect(subject).to eq(false)
+  end
+
+  it "returns false if the the item save is false" do
+    expect(item).to receive(:save).and_return(false)
```
Comment (jgondron) [inline spec/services/save_metadata_spec.rb]: Should just be an allow.

File: spec/controllers/v1/metadata_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  let(:collection) { instance_double(Collection, id: "1", updated_at: nil, items: nil, collection_configuration: collection_configuration) }
+  let(:item) { instance_double(Item, id: "1", collection: collection, children: nil) }
+
+  before(:each) do
+    allow_any_instance_of(ItemQuery).to receive(:public_find).and_return(item)
+    allow_any_instance_of(CollectionQuery).to receive(:public_find).and_return(collection)
+  end
+
+  describe "PUT #update" do
+    let(:collection_configuration) { CollectionConfiguration.new }
```
Comment (rfox2) [inline spec/controllers/v1/metadata_controller_spec.rb]: Why is this needed when you have :collection_configuration assigned above?

File: spec/controllers/v1/metadata_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+      expect(response).to be_unprocessable
+    end
+
+    it "renders with errors" do
+      allow(SaveMetadata).to receive(:call).and_return(false)
+      subject
+      expect(response).to render_template("errors")
+    end
+
+    it "assigns and item" do
```
Comment (rfox2) [inline spec/controllers/v1/metadata_controller_spec.rb]: Should this be "assigns an item"?

File: app/services/save_metadata.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+  def initialize(item, params)
+    @params = params
+    @item = item
+  end
+
+  def save
+    fix_params
+    set_metadata
+    if valid? && item.save
```
Comment (rfox2) [inline app/services/save_metadata.rb]: Should there be a spec to test for valid metadata? i.e. - a spec that tests that false gets returned if the metadata is invalid, and a spec that tests to see if the metadata is returned on success?

File: spec/controllers/v1/metadata_controller_spec.rb (hunk 7 above)
Comment (jonhartzler) [inline spec/controllers/v1/metadata_controller_spec.rb]: I removed the unused let statements from above.

**Comment by jgondron** [comment]
👍

**Comment by rfox2** [comment]
👍
